### PR TITLE
doc(pubsub): delete schemas > 48 hours old

### PR DESCRIPTION
**Pull request by alevenberg**
Issue #11720

- - -
This change is [Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11862)

### CODE REVIEW

File: google/cloud/pubsub/samples/pubsub_samples_common.cc [hunk 1]
```diff
@@ -188,6 +189,23 @@ std::string CommitSchemaRevisionsForRollbackSchemaTesting(
...
+  for (auto& schema : schema_admin.ListSchemas(parent)) {
+    if (!schema) continue;
+    absl::Time schema_create_time =
+        absl::FromUnixSeconds((*schema).revision_create_time().seconds()) +
+        absl::Nanoseconds((*schema).revision_create_time().nanos());
+
+    if (schema_create_time < time_now - absl::Hours(48)) {
+      google::pubsub::v1::DeleteSchemaRequest request;
+      request.set_name((*schema).name());
+      schema_admin.DeleteSchema(request);
```
Comment (coryan) [inline google/cloud/pubsub/samples/pubsub_samples_common.cc:204]: nit: we usually write `(void)schema_admin.DeleteSchema(request)` to signal that we really want to ignore the return value.

File: google/cloud/pubsub/samples/samples.cc [hunk 2]
```diff
@@ -2123,6 +2124,8 @@ void AutoRunAvro(
 
   std::cout << "\nRunning DeleteSchema() sample [avro]" << std::endl;
   DeleteSchema(schema_admin, {project_id, avro_schema_id});
+
+  CleanupSchemas(schema_admin, project_id, absl::Now());
```
Comment (coryan) [inline google/cloud/pubsub/samples/samples.cc:2128]: Please move this to the beginning of the function.  We want this to run first so we can recover quota space, if we ever run out, before starting the test.

File: google/cloud/pubsub/samples/pubsub_samples_common.cc [hunk 3]
```diff
@@ -188,6 +189,23 @@ std::string CommitSchemaRevisionsForRollbackSchemaTesting(
...
 }
 
+void CleanupSchemas(google::cloud::pubsub::SchemaServiceClient& schema_admin,
+                    std::string const& project_id, absl::Time const& time_now) {
+  auto const parent = google::cloud::Project(project_id).FullName();
+  for (auto& schema : schema_admin.ListSchemas(parent)) {
+    if (!schema) continue;
+    absl::Time schema_create_time =
+        absl::FromUnixSeconds((*schema).revision_create_time().seconds()) +
+        absl::Nanoseconds((*schema).revision_create_time().nanos());
```
Comment (coryan) [inline google/cloud/pubsub/samples/pubsub_samples_common.cc:199]: We have a helper:

https://github.com/googleapis/google-cloud-cpp/blob/654439aadc22fae0e8fcf8a030fca80d860c7182/google/cloud/internal/time_utils.h#L42

consider:

```suggestion
auto const schema_create_time =
    google::cloud::internal::ToAbslTime(schema->revision_create_time());
```

File: google/cloud/pubsub/samples/pubsub_samples_common.cc [hunk 4]
```diff
@@ -188,6 +189,23 @@ std::string CommitSchemaRevisionsForRollbackSchemaTesting(
...
+
+    if (schema_create_time < time_now - absl::Hours(48)) {
+      google::pubsub::v1::DeleteSchemaRequest request;
+      request.set_name((*schema).name());
```
Comment (coryan) [inline google/cloud/pubsub/samples/pubsub_samples_common.cc:203]: nit: we spell that: `schema->name()`

**Review by coryan — CHANGES_REQUESTED** [review]

File: google/cloud/pubsub/samples/pubsub_samples_common.cc [hunk 5]
```diff
@@ -188,6 +189,23 @@ std::string CommitSchemaRevisionsForRollbackSchemaTesting(
...
+void CleanupSchemas(google::cloud::pubsub::SchemaServiceClient& schema_admin,
+                    std::string const& project_id, absl::Time const& time_now) {
+  auto const parent = google::cloud::Project(project_id).FullName();
+  for (auto& schema : schema_admin.ListSchemas(parent)) {
+    if (!schema) continue;
+    absl::Time schema_create_time =
+        absl::FromUnixSeconds((*schema).revision_create_time().seconds()) +
+        absl::Nanoseconds((*schema).revision_create_time().nanos());
+
+    if (schema_create_time < time_now - absl::Hours(48)) {
```
Comment (devbww) [inline google/cloud/pubsub/samples/pubsub_samples_common.cc:201]: Should we also only be considering schemas with the `RandomSchemaId()` prefix?

File: google/cloud/pubsub/samples/samples.cc [hunk 6]
```diff
@@ -2125,7 +2126,8 @@ void AutoRunAvro(
   std::cout << "\nRunning DeleteSchema() sample [avro]" << std::endl;
   DeleteSchema(schema_admin, {project_id, avro_schema_id});
 
-  CleanupSchemas(schema_admin, project_id, absl::Now());
+  std::chrono::system_clock clock;
+  CleanupSchemas(schema_admin, project_id, absl::FromChrono(clock.now()));
```
Comment (coryan) [inline google/cloud/pubsub/samples/samples.cc:2130]: nit-ish: this is a static function, you can say `std::chrono::system_clock::now()`

**Review by coryan — APPROVED** [review]

**Review by alevenberg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11862)* status: 0 of 3 files reviewed, 4 unresolved discussions (waiting on @coryan and @devbww)

___
*[`google/cloud/pubsub/samples/pubsub_samples_common.cc` line 199 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11862#gh-1226873686:-NXkfldu7WN-DaaHIK5R:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ec69aa5be3c12c1dfe091e36860888b940df9897/google/cloud/pubsub/samples/pubsub_samples_common.cc#L199)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

We have a helper:

https://github.com/googleapis/google-cloud-cpp/blob/654439aadc22fae0e8fcf8a030fca80d860c7182/google/cloud/internal/time_utils.h#L42

consider:

```suggestion
auto const schema_create_time =
    google::cloud::internal::ToAbslTime(schema->revision_create_time());
```
</blockquote></details>

Done.
___
*[`google/cloud/pubsub/samples/pubsub_samples_common.cc` line 201 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11862#gh-1227026770:-NXkg8P5-uuliEBKwvqb:b-rt8mpi) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ec69aa5be3c12c1dfe091e36860888b940df9897/google/cloud/pubsub/samples/pubsub_samples_common.cc#L201)):*
<details><summary><i>Previously, devbww (Bradley White) wrote…</i></summary><blockquote>

Should we also only be considering schemas with the `RandomSchemaId()` prefix?
</blockquote></details>

That's a good idea. Added a check.
___
*[`google/cloud/pubsub/samples/samples.cc` line 2128 at r1](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11862#gh-1226870444:-NXki2GR29Jur-sxoqwS:b-896fix) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/ec69aa5be3c12c1dfe091e36860888b940df9897/google/cloud/pubsub/samples/samples.cc#L2128)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

Please move this to the beginning of the function.  We want this to run first so we can recover quota space, if we ever run out, before starting the test.
</blockquote></details>

Done.
___
*[`google/cloud/pubsub/samples/samples.cc` line 2130 at r2](https://reviewable.io/reviews/googleapis/google-cloud-cpp/11862#gh-1227029792:-NXkhRQA4y15_ugd2On6:b4q3bpt) ([raw file](https://github.com/googleapis/google-cloud-cpp/blob/645bfc253e256e46fa8e23d77793ea412216c3f5/google/cloud/pubsub/samples/samples.cc#L2130)):*
<details><summary><i>Previously, coryan (Carlos O'Ryan) wrote…</i></summary><blockquote>

nit-ish: this is a static function, you can say `std::chrono::system_clock::now()`
</blockquote></details>

The old cmake build didn't work with absl::Now as well :(

Fixed.

**Pull request merged**